### PR TITLE
rgw_file: fix write error when the write offset overlaps

### PR DESCRIPTION
**Pull request by yaozongyou**
fix bug in http://tracker.ceph.com/issues/21455

Signed-off-by: Yao Zongyou <yaozongyou@vip.qq.com>

### CODE REVIEW

**Comment by yaozongyou** [comment]
The nfs client writes data to system's page cache, and the page cache will be flushed to rgw_file by nfs-ganesha. The page cache is page size aligned, so the write offset may overlap in a page size.

**Comment by yaozongyou** [comment]
@tchaikov, would you like to have a look at this?

**Comment by tchaikov** [comment]
@yaozongyou i am not familiar with rgw, thus not qualified to review your change.

**Comment by mattbenjamin** [comment]
@yaozongyou does this change attempt to address a behavior seen in Linux clients mounting -osync?

**Comment by yaozongyou** [comment]
image
yes, as above image shows, even mounted using -osync, the offset may overlap between two requests.

**Comment by mattbenjamin** [comment]
@yaozongyou do you have a simple reproducer from the client side?

File: src/rgw/rgw_file.cc [hunk 1]
```diff
@@ -1159,6 +1159,15 @@ namespace rgw {
       }
     }
 
+    int overlap = 0;
+    if ((static_cast<off_t>(off) < f->write_req->real_ofs) &&
+        (static_cast<long>(f->write_req->real_ofs - off) < sysconf(_SC_PAGESIZE))) {
```
Comment (mattbenjamin) [inline src/rgw/rgw_file.cc]: of course, the client's pagesize (like its mount options) is not knowable, right?  if we unconditionally attempt this correction, do we do just as well?

**Comment by yaozongyou** [comment]
@mattbenjamin a simple reproducer from the client side:
https://gist.github.com/yaozongyou/5f5ed0d362ccef97942a09496f4e2a8c

**Comment by yaozongyou** [comment]
@mattbenjamin can you reproduce it in your environment?

**Comment by mattbenjamin** [comment]
@yaozongyou I'll attempt it--if my results are consistent, this should be ok (with the possible caveat about pagesize flagged above)

**Comment by mattbenjamin** [comment]
@yaozongyou this doesn't error in my environment :(

**Comment by mattbenjamin** [comment]
@yaozongyou so, we've discussed this here a bit;  it's known that there may be kernels/clients that don't conform to our expectation that WRITEs will be strictly sequential even when a) the mount is -osync and b) the application i/o is strictly sequential--but the kernels we've explicitly tested do so.  I'd be interested to know which distro and kernel version you're doing this, on what architecture.  My inclination is not to apply this or previous suggestions like this, but rather to continue with our work in progress which will allow async and random i/o.  Please note that iiuc, your posix write calls are not guaranteed to write all bytes in one invocation--you need to be parepared for short writes.  In any case, I will expedite getting that finished up, becausel clearly it is an issue for users.

Matt

**Comment by yaozongyou** [comment]
@mattbenjamin We are using CentOS Linux release 7.3.1611, the kernel version is 3.10.0-514.el7.x86_64.
We have a script using the system command `ftp` downloading files from remote ftp server directly into local mounted rgw buckets. The ftp command gives error. After stracing, we make a simple reproducer as above.
We mount using mount -t nfs -o sync 127.0.0.1:/ /mnt and mount -t nfs  -osync,vers=4.1  127.0.0.1:/ /mnt. Both of them didn't work.
In /etc/ganesha/ganesha.conf , the Path is the bucket name. the command showmount -e shows as follows:

Export list for localhost:
testbucket (everyone)

**Comment by yaozongyou** [comment]
@mattbenjamin I have made some changes about the question you mentioned. And I have tested in our environment, it works very well.

**Pull request merged**